File: core/src/main/java/com/pierfrancescosoffritti/androidyoutubeplayer/utils/YouTubePlayerTracker.java
package com.pierfrancescosoffritti.androidyoutubeplayer.utils;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.pierfrancescosoffritti.androidyoutubeplayer.player.PlayerConstants;
import com.pierfrancescosoffritti.androidyoutubeplayer.player.listeners.AbstractYouTubePlayerListener;

/**
 * Utility class responsible for tracking the state of YouTubePlayer.
 * This is a YouTubePlayerListener, therefore is responsibility of the user to add and remove it as a listener on the YouTubePlayer object.
 */
public class YouTubePlayerTracker extends AbstractYouTubePlayerListener {
    private PlayerConstants.PlayerState currentState = PlayerConstants.PlayerState.UNKNOWN;
    private float currentSecond;
    private float videoDuration;
    private String videoId;

    @Override
    public void onStateChange(@NonNull PlayerConstants.PlayerState state) {
        this.currentState = state;
    }

    @Override
    public void onCurrentSecond(float second) {
        currentSecond = second;
    }

    @Override
    public void onVideoDuration(float duration) {
        videoDuration = duration;
    }

    @Override
    public void onVideoId(@NonNull String videoId) {
        this.videoId = videoId;
    }

    /**
     * @return the player state. A value from {@link PlayerConstants.PlayerState}
     */
    public PlayerConstants.PlayerState getState() {
        return currentState;
    }

    public float getCurrentSecond() {
        return currentSecond;
    }

    public float getVideoDuration() {
        return videoDuration;
    }

    public @Nullable String getVideoId() {
        return videoId;
    }
}

File: core/src/main/java/com/pierfrancescosoffritti/androidyoutubeplayer/core/player/utils/YouTubePlayerTracker.kt
package com.pierfrancescosoffritti.androidyoutubeplayer.core.player.utils

import com.pierfrancescosoffritti.androidyoutubeplayer.core.player.PlayerConstants
import com.pierfrancescosoffritti.androidyoutubeplayer.core.player.YouTubePlayer
import com.pierfrancescosoffritti.androidyoutubeplayer.core.player.listeners.AbstractYouTubePlayerListener

/**
 * Utility class responsible for tracking the state of YouTubePlayer.
 * This is a YouTubePlayerListener, therefore to work it has to be added as listener to a YouTubePlayer.
 */
class YouTubePlayerTracker : AbstractYouTubePlayerListener() {
  /**
   * @return the player state. A value from [PlayerConstants.PlayerState]
   */
  var state: PlayerConstants.PlayerState = PlayerConstants.PlayerState.UNKNOWN
    private set
  var currentSecond: Float = 0f
    private set
  var videoDuration: Float = 0f
    private set
  var videoId: String? = null
    private set

  override fun onStateChange(youTubePlayer: YouTubePlayer, state: PlayerConstants.PlayerState) {
    this.state = state
  }

  override fun onCurrentSecond(youTubePlayer: YouTubePlayer, second: Float) {
    currentSecond = second
  }

  override fun onVideoDuration(youTubePlayer: YouTubePlayer, duration: Float) {
    videoDuration = duration
  }

  override fun onVideoId(youTubePlayer: YouTubePlayer, videoId: String) {
    this.videoId = videoId
  }
}
